modules/dcache/src/main/java/org/dcache/pool/statistics/IoStatistics.java

```diff
@@ -39,15 +41,16 @@
      *         The duration of the read request in nanoseconds; must be non-negative
      */
 
-    public void updateRead(long readBytes, long readTime){
+    public void updateRead(long readBytes, long readTime) throws ArrayIndexOutOfBoundsException{
 
         double readSpeed = calculateSpeed(readBytes, readTime);
-        _readSpeeds.add(readSpeed);
+        _readSpeeds[_readSpeedsArrayIndex] = readSpeed;
```

Comment:
what happens when the array index >= 8192?

I think you need to make this a circular buffer.  The percentile computation then is valid only over the last 8192 values.

You would need to do something likeL

_readSpeedsArrayIndex = (_readSpeedsArrayIndex + 1) % _readSpeeds.len;

to find the index of the next value to be inserted.